### DISCUSSION TITLE
[appDir] Instant loading state, renders UI of current or next route?

### DISCUSSION THREAD

**Discussion by kuus** [vercel/next.js#47157] [Help] [answered]
### Summary

Hello, maybe I am missing some fundamentals of the Next 13 [Instant loading state](https://beta.nextjs.org/docs/routing/loading-ui#instant-loading-states). 
I initially thought that this feature purpose was to render a skeleton of the page the user is going to _before_ that was ready. For instance, let's take this folder structure:

```bash
/app/
|__ layout.tsx # with shared Header and Footer
|__ page.tsx # home page
|__ loading.tsx # A: just a centered spinner (we do not know what is going to be within here)
|__ product/[id]/
      |__ page.tsx
      |__ loading.tsx # B: a precise skeleton of the single product page
 |__ category/[id]/
      |__ page.tsx
      |__ loading.tsx # C: a precise skeleton of the single category page
```
On the `home` page we have two links, one to a featured `product` and one to a featured `category`. The three pages only share the same Header and Footer in their common _layout_. 
Now on the`home` page a user clicks on a featured `product`. I would expect the `loading.tsx B` to render immediately.
And if the user would have clicked on the featured `category` I would expect  the `loading.tsx C` to render immediately instead.
It seems that in either case `loading.tsx A` is rendered instead. But in this loading template there is no UI helping the transition to the UI the user is going to, which renders the _Instant loading states_ not suited for links between "unrelated" pages.

is this by design?
or am I missing something basic?

thx!

### Additional information

As a next 12 user I still feel [this somewhat related issue](https://github.com/vercel/next.js/discussions/23992) is what hurts my websites UX the most, so I hoped thew new Instant loading states would have make a big difference and represetn a good enough reason to migrate to next 13. But given the above scenario I am a bit skeptical now as that kind of loading state only aware of where the user is and not of where the user is going was already possible using the next 12 router's events.

@leerob maybe can you check this and share the design idea of the nextjs team in regards to this? thx!

**Comment by leerob** [comment, marked as answer]
`loading.js` will be nested inside `layout.js`. It'll wrap the `page.js` file and any children below in a `<Suspense>` boundary. So, if you want nested `loading.js` files, you would need a layout for products and categories. Does that make sense? You can also use `<Suspense>` directly.

**Reply by kuus** [reply]
I still haven't had a chance to try it but it makes sense, especially using Suspense directly, thanks @leerob

**Reply by fprl** [reply]
Hey there! I'm having this problem where the instant loading navigation is not working.

Example: https://covered-homes.vercel.app/

I have this structure:
```
/app/
|__ layout.tsx // includes header
|__ page.tsx // home page
|__ (booking)
      |__ layout.tsx // includes search-bar (filters)
      |__ huis-ibiza/
            |__ page.tsx
            |__ loading.tsx // skeleton for page
```

When navigating from `/` to `/huis-ibiza` you can see that the navigation is not instant, and it awaits for the result of the promise (slow API).

How does my API call look?
```
return fetch(
    `${apis.v4}/house/search/${locale}?page=${pageNr}&results=${resultsNr}&${filters}`,
    {
      cache: 'no-store',
      next: {
        revalidate: 0,
      },
    }
  );
```


### Summary
1. navigation is not instant, and loading.tsx shows for a second.
2. when navigating from `/huis-ibiza` ---> `/` ---> `/huis-ibiza`:
    a. loading.tsx not visible.
    b. navigation and results are instantaneous, so it's using cache, although I specified to be dynamic in the fetch function.
